symbolic-base/src/ZkFold/Symbolic/Data/Int.hs

```diff
@@ -0,0 +1,216 @@
+import           Test.QuickCheck                  (Arbitrary (..))
+
+import           ZkFold.Base.Algebra.Basic.Class  hiding (Euclidean (..))
+import           ZkFold.Base.Algebra.Basic.Number
+import           ZkFold.Base.Data.Vector          (Vector (..))
+import           ZkFold.Symbolic.Class
+import           ZkFold.Symbolic.Data.Bool
+import           ZkFold.Symbolic.Data.Class       (SymbolicData)
+import           ZkFold.Symbolic.Data.Combinators
+import           ZkFold.Symbolic.Data.Conditional
+import           ZkFold.Symbolic.Data.Eq
+import           ZkFold.Symbolic.Data.Ord
+import           ZkFold.Symbolic.Data.UInt
+import           ZkFold.Symbolic.Interpreter      (Interpreter (..))
+
+
+newtype Int (n :: Natural) (r :: RegisterSize) (c :: (Type -> Type) -> Type) = Int { uint :: UInt (n+1) r c}
+
+deriving instance Generic (Int n r c)
+deriving instance (NFData (c (Vector (NumberOfRegisters (BaseField c) (n+1) r)))) => NFData (Int n r c)
+deriving instance (Haskell.Eq (c (Vector (NumberOfRegisters (BaseField c) (n+1) r)))) => Haskell.Eq (Int n r c)
+deriving instance (Haskell.Show (BaseField c), Haskell.Show (c (Vector (NumberOfRegisters (BaseField c) (n+1) r))))
+    => Haskell.Show (Int n r c)
+deriving instance (KnownRegisters c (n+1) r, Symbolic c) => SymbolicData (Int n r c)
+deriving instance (KnownRegisters c (n+1) r, Symbolic c) => Conditional (Bool c) (Int n r c)
+deriving instance (KnownRegisters c (n+1) r, Symbolic c) => Eq (Int n r c)
+
+instance (Symbolic c, KnownNat n, KnownRegisterSize r) => FromConstant Natural (Int n r c) where
+    fromConstant c = withDict (plusNat @n @1) $ Int (fromConstant c)
+
+instance (Symbolic c, KnownNat n, KnownRegisterSize r) => FromConstant Integer (Int n r c) where
+    fromConstant c = withDict (plusNat @n @1) $ Int (Haskell.bool (fromConstant c) (negate $ fromConstant (Haskell.negate c)) (c Haskell.<0))
+
+instance MultiplicativeMonoid (Int n r c) => Exponent (Int n r c) Natural where
+    (^) = natPow
+
+------------------------------------------------------------------------------
+
+instance
+    ( Symbolic (Interpreter a), KnownNat n, KnownRegisterSize r
+    , KnownNat (n + 1)
+    ) => ToConstant (Int n r (Interpreter a)) where
+    type Const (Int n r (Interpreter a)) = Integer
+    toConstant i@(Int u) =
+        withDict (plusNat @n @1) $
+            withGetRegisterSize @(n+1) @r @a $
+                Haskell.bool (Haskell.toInteger $ toConstant u) (negate . Haskell.toInteger . toConstant $ negate u) (isNegative i Haskell.== true)
+
+isNegative :: forall n r c.
+    ( Symbolic c, KnownNat n, KnownRegisterSize r
+    ) => Int n r c -> Bool c
+isNegative (Int u) =
+    withDict (plusNat @n @1) $
+        withNumberOfRegisters @(n+1) @r @(BaseField c) $
+            withGetRegisterSize @(n+1) @r @(BaseField c) $
+                withCeilRegSize @(GetRegisterSize (BaseField c) (n+1) r) @OrdWord $
+                    u >= fromConstant ((2 :: Natural) ^ (value @n))
```

Comment:
This should optimize away to taking the highest bit in the representation. It would be clearer if we explicitly did this here.